### ISSUE TITLE
Validate Docker v1.11

### ISSUE THREAD

**Issue by vishh** [kubernetes/kubernetes#23397]
Docker v1.11 rcs are starting to show up. Its time to get started with the validation.

https://github.com/docker/docker/releases/tag/v1.11.0-rc1

cc @kubernetes/sig-node 

EDIT(timstclair):

TASKS:
- [x] e2e tests pass
- [x] performance analysis
- [x] startup tests (bootstrap, restart)
- [ ] live upgrade successful
- [x] 1 week soak tests

**Comment by timstclair** [comment]
xref https://github.com/kubernetes/kubernetes/issues/22765 - automate this process

**Comment by luxas** [comment]
The really big change here is that you need four binaries instead of one: `docker`, `runc`, `containerd`, `containerd-shim` if I recall correctly.
That _may_ cause issues

**Comment by vishh** [comment]
The salt script uses packages. So multiple binaries should not be an issue
I think.

**Comment by yujuhong** [comment]
We may want to put both containerd and docker into one cgroup to account for the resource usage.

**Comment by timstclair** [comment]
kube-proxy fails to come up. From `/var/log/kube-proxy.log`
```
Flag --resource-container has been deprecated, This feature will be removed in a later release.
I0324 23:56:52.790786       6 iptables.go:177] Could not connect to D-Bus system bus: dial unix /var/run/dbus/system_bus_socket: no such file or directory
I0324 23:56:52.790882       6 server.go:154] setting OOM scores is unsupported in this build
I0324 23:56:52.855289       6 server.go:200] Using iptables Proxier.
F0324 23:56:52.855424       6 server.go:208] Unable to create proxier: can't set sysctl net/ipv4/conf/all/route_localnet: open /proc/sys/net/ipv4/conf/all/route_localnet: read-only file system
...
```

**Comment by yujuhong** [comment]
runc by default mounts `/proc/sys` as read-only (https://github.com/opencontainers/runc/blob/24950964ec70f18d18e767e9141d8d952efa416b/utils.go#L131)

The privileged container will need to unmount it to write to it. I added it to the manifest and kube-proxy seems to be working after that.

**Comment by vishh** [comment]
https://github.com/docker/docker/releases/tag/v1.11.0-rc2

**Comment by timstclair** [comment]
Results from running e2e tests on v1.11.0-rc2:

[kubernetes-e2e-gce](https://gist.github.com/timstclair/3a000e1a9e22c9d204397fa3d25e7f43)
```
Summarizing 6 Failures:

[Fail] [k8s.io] Addon update [BeforeEach] should propagate add-on file changes 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/addon_update.go:208

[Fail] [k8s.io] Horizontal pod autoscaling (scale resource: CPU) [k8s.io] ReplicationController light [It] Should scale from 1 pod to 2 pods 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/autoscaling_utils.go:325

[Fail] [k8s.io] Services [It] should be able to up and down services 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:248

[Fail] [k8s.io] SSH [It] should SSH to all nodes and run commands 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/ssh.go:65

[Fail] [k8s.io] Horizontal pod autoscaling (scale resource: CPU) [k8s.io] ReplicationController light [It] Should scale from 2 pods to 1 pod using HPA version v1 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/autoscaling_utils.go:325

[Fail] [k8s.io] Monitoring [It] should verify monitoring pods and all cluster nodes are available on influxdb using heapster. 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/util.go:1591

Ran 161 of 268 Specs in 5971.947 seconds
FAIL! -- 155 Passed | 6 Failed | 0 Pending | 107 Skipped --- FAIL: TestE2E (5972.33s)
FAIL
```

[kubernetes-e2e-gce-serial](https://gist.github.com/timstclair/d81d3be30eb8d8444796acbf59dfecf4)
```
Summarizing 18 Failures:

[Fail] [k8s.io] Services [It] should work after restarting apiserver [Disruptive] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:349

[Fail] [k8s.io] Nodes [Disruptive] [k8s.io] Network when a node becomes unreachable [It] All pods on the unreachable node should be marked as NotReady upon the node turn NotReady AND all pods should be mark back to Ready when the node get back to Ready before pod eviction timeout 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/util.go:3549

[Fail] [k8s.io] Horizontal pod autoscaling (scale resource: CPU) [k8s.io] [Serial] [Slow] ReplicationController [It] Should scale from 5 pods to 3 pods and from 3 to 1 and verify decision stability 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/autoscaling_utils.go:325

[Fail] [k8s.io] Etcd failure [Disruptive] [It] should recover from network partition with master 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/etcd_failure.go:93

[Fail] [k8s.io] DaemonRestart [Disruptive] [It] Kubelet should not restart containers across restart 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/daemon_restart.go:62

[Fail] [k8s.io] Horizontal pod autoscaling (scale resource: CPU) [k8s.io] [Serial] [Slow] Deployment [It] Should scale from 5 pods to 3 pods and from 3 to 1 and verify decision stability 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/autoscaling_utils.go:325

[Fail] [k8s.io] Etcd failure [Disruptive] [It] should recover from SIGKILL 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/etcd_failure.go:93

[Fail] [k8s.io] Services [It] should work after restarting kube-proxy [Disruptive] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/service.go:307

[Fail] [k8s.io] Nodes [Disruptive] [k8s.io] Network when a node becomes unreachable [It] [replication controller] recreates pods scheduled on the unreachable node AND allows scheduling of pods on a node after it rejoins the cluster 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/util.go:3549

[Fail] [k8s.io] SchedulerPredicates [Serial] [It] validates MaxPods limit number of pods that are allowed to run [Slow] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/util.go:2159

[Fail] [k8s.io] Horizontal pod autoscaling (scale resource: CPU) [k8s.io] [Serial] [Slow] ReplicationController [It] Should scale from 1 pod to 3 pods and from 3 to 5 and verify decision stability 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/autoscaling_utils.go:325

[Fail] [k8s.io] Horizontal pod autoscaling (scale resource: CPU) [k8s.io] [Serial] [Slow] Deployment [It] Should scale from 1 pod to 3 pods and from 3 to 5 and verify decision stability 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/autoscaling_utils.go:325

[Fail] [k8s.io] Horizontal pod autoscaling (scale resource: CPU) [k8s.io] [Serial] [Slow] ReplicaSet [It] Should scale from 1 pod to 3 pods and from 3 to 5 and verify decision stability 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/autoscaling_utils.go:325

[Fail] [k8s.io] Kubelet [Serial] [Slow] [k8s.io] regular resource usage tracking [It] for 100 pods per node over 20m0s 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubelet_perf.go:72

[Fail] [k8s.io] Kubelet [Serial] [Slow] [k8s.io] regular resource usage tracking [It] for 35 pods per node over 20m0s 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubelet_perf.go:143

[Fail] [k8s.io] DaemonRestart [Disruptive] [It] Controller Manager should not create/delete replicas across restart 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/daemon_restart.go:62

[Fail] [k8s.io] Horizontal pod autoscaling (scale resource: CPU) [k8s.io] [Serial] [Slow] ReplicaSet [It] Should scale from 5 pods to 3 pods and from 3 to 1 and verify decision stability 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/autoscaling_utils.go:325

[Fail] [k8s.io] DaemonRestart [Disruptive] [It] Scheduler should continue assigning pods to nodes across restart 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/daemon_restart.go:62

Ran 30 of 268 Specs in 9137.840 seconds
FAIL! -- 12 Passed | 18 Failed | 0 Pending | 238 Skipped --- FAIL: TestE2E (9138.18s)
FAIL
```

[kubernetes-e2e-gce-slow](https://gist.github.com/timstclair/6125760a05654d36c255048c49ef7ea5)
```
SUCCESS! -- 9 Passed | 0 Failed | 0 Pending | 259 Skipped PASS
```

Lots to look into here...

**Comment by vishh** [comment]
FYI: RC4 is out - Linux 64bit tgz: https://test.docker.com/builds/Linux/x86_64/docker-1.11.0-rc4.tgz

**Comment by timstclair** [comment]
Blocker: https://github.com/google/cadvisor/issues/1206

**Comment by yujuhong** [comment]
FYI, `docker-containerd` should probably be moved to the same cgroup that docker daemon is in so that we can account the resource overhead of containerd-shim, a per-container process.  The containerd-shim uses about 3-5MB of memory and may be rewritten in C (xref: https://github.com/docker/docker/issues/21737)

**Comment by timstclair** [comment]
FYI, docker v1.11 is [released](https://blog.docker.com/2016/04/docker-engine-1-11-runc/)

**Comment by yujuhong** [comment]
> runc by default mounts /proc/sys as read-only (https://github.com/opencontainers/runc/blob/24950964ec70f18d18e767e9141d8d952efa416b/utils.go#L131)
> 
> The privileged container will need to unmount it to write to it. I added it to the manifest and kube-proxy seems to be working after that.
> 
> EDIT: I added umount /proc/sys as part of the command in the manifest file. The pod was up and running after that (i.e., not crashing anymore).

By the way, this should was already fixed by https://github.com/docker/docker/pull/21751

**Comment by timstclair** [comment]
Almost all tests passing, yay! Remaining failures are:
```
[Fail] [k8s.io] Kubelet [Serial] [Slow] [k8s.io] regular resource usage tracking [It] resource tracking for 35 pods per node 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubelet_perf.go:150

[Fail] [k8s.io] SchedulerPredicates [Serial] [It] validates MaxPods limit number of pods that are allowed to run [Slow] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2091

[Fail] [k8s.io] Kubelet [Serial] [Slow] [k8s.io] regular resource usage tracking [It] resource tracking for 100 pods per node 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubelet_perf.go:74

[Fail] [k8s.io] Kubelet [Serial] [Slow] [k8s.io] regular resource usage tracking [It] resource tracking for 0 pods per node 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubelet_perf.go:150
```

3 of those are expected increases in memory usage (as described [here](https://github.com/kubernetes/kubernetes/issues/23397#issuecomment-209583923)). I think the other issue is just that my test cluster has insufficient disk quota (fails on master as well).

**Comment by yujuhong** [comment]
> Almost all tests passing, yay!

Awesome! 

> 3 of those are expected increases in memory usage (as described here). I think the other issue is just that my test cluster has insufficient disk quota (fails on master as well).

Just wanted to double-check: Is containerd in the same cgroup with docker? If so, we can check the memory increase and bump the limits in the test accordingly.

**Comment by timstclair** [comment]
Performed manual restart tests, no issues found.

Tested the following with no apparent issues:
- Kill docker daemon (`sudo pkill -9 -f /usr/bin/dock[er]`) while running steady state with lots of pause pods, wait and restart (`sudo /etc/init.d/docker start`)
- Kill docker daemon while scaling pause (gcr.io/google_containers/pause) replication controller to 0 replicas. `kubectl` reports pods as terminated. Restart docker daemon, containers are quickly terminated.
- Kill docker daemon while scaling pause rc to 200 replicas (3 node + 1 master cluster). Pods eventually report error status. Restart docker daemon, pods eventually report running.
- All of the above on the master node
- All of the above killing docker-containerd (it restarts automatically, with some backoff)

**Comment by timstclair** [comment]
_v1.9 results copied from [this comment](https://github.com/kubernetes/kubernetes/issues/16110#issuecomment-180510177)_

**Introduction**
I benchmarked docker v1.11 using the [docker micro benchmark tool](https://github.com/kubernetes/contrib/tree/master/docker-micro-benchmark).

**Benchmark Environment**
* Cloud Provider: GCE
* VM Instance: n1-standard-2 (2 vCPUs, 7.5 GB memory)
* OS: Debian GNU/Linux 8.3 (jessie)
* Docker version:
```
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:11:07 2016
 OS/Arch:      linux/amd64
```

### 1. Benchmark list/inspect with varying number of containers
Using `benchmark.sh -c`

- **Latency**
  - Docker 1.9.1
![latency-varies_container](https://cloud.githubusercontent.com/assets/5821883/12855505/bc233386-cbf4-11e5-9a6d-c83749d733f6.png)
  - Docker 1.11.1
![latency-varies_container](https://cloud.githubusercontent.com/assets/3272040/15033929/10b4783a-1225-11e6-8a77-6e2eb3d697e9.png)
- **CPU Usage**
 - Docker 1.9.1
![cpu](https://cloud.githubusercontent.com/assets/5821883/12855479/8a96cd96-cbf4-11e5-8e2e-26526c357ff4.png)
  - Docker 1.11.1
![cpu](https://cloud.githubusercontent.com/assets/3272040/15033931/1b998786-1225-11e6-8e4f-f1c8a418e9ee.png)

### 2. Benchmark list/inspect with varying operation intervals
Using `benchmark.sh -i`

* **`ps` [all=true] Latency**
 * Docker 1.9.1
![latency-list_all](https://cloud.githubusercontent.com/assets/5821883/12855829/84f42a76-cbf6-11e5-8864-1b90a89bbaea.png)
 * Docker 1.11.1
![latency-list_all](https://cloud.githubusercontent.com/assets/3272040/15033938/2e22add8-1225-11e6-938f-d4f0209dae77.png)

* **`ps` [all=false] Latency**
 * Docker 1.9.1
![latency-list_alive](https://cloud.githubusercontent.com/assets/5821883/12855835/90d49a38-cbf6-11e5-84e9-62934070f4f2.png)
 * Docker 1.11.1
![latency-list_alive](https://cloud.githubusercontent.com/assets/3272040/15033941/3680bff6-1225-11e6-8433-676d19611c5c.png)

* **`inspect` Latency**
 * Docker 1.9.1
![latency-inspect](https://cloud.githubusercontent.com/assets/5821883/12855869/c2b81520-cbf6-11e5-83a9-159f060cbe79.png)
 * Docker 1.11.1
![latency-inspect](https://cloud.githubusercontent.com/assets/3272040/15033944/3dcc0658-1225-11e6-84d7-c8b96e85b991.png)

* **CPU Usage**
 * Docker 1.9.1
![cpu](https://cloud.githubusercontent.com/assets/5821883/12855878/d34510e6-cbf6-11e5-81a5-0fb83bcd721f.png)
 * Docker 1.11.1
![cpu](https://cloud.githubusercontent.com/assets/3272040/15033946/42124e20-1225-11e6-9065-3f6acece7a82.png)

### 3. Benchmark list/inspect with varying number of goroutines
Using `benchmark.sh -r`

* **`ps` [all=true] Latency**
 * Docker 1.9.1
![latency-list_all](https://cloud.githubusercontent.com/assets/5821883/12856118/203b6ec6-cbf8-11e5-896a-747e16f85669.png)
 * Docker 1.11.1
![latency-list_all](https://cloud.githubusercontent.com/assets/3272040/15033962/8536f49e-1225-11e6-8d61-94a2d717895b.png)

* **`inspect` Latency**
 * Docker 1.9.1
![latency-inspect](https://cloud.githubusercontent.com/assets/5821883/12856134/36111da4-cbf8-11e5-9fae-990189e2e809.png)
 * Docker 1.11.1
![latency-inspect](https://cloud.githubusercontent.com/assets/3272040/15033963/8a438a60-1225-11e6-8797-4f656f8b064c.png)

* **CPU Usage**
 * Docker 1.9.1
![cpu](https://cloud.githubusercontent.com/assets/5821883/12856172/6374ffd6-cbf8-11e5-9f99-acd8938efd52.png)
 * Docker 1.11.1
![cpu](https://cloud.githubusercontent.com/assets/3272040/15033966/8f1ed8a0-1225-11e6-93ef-c2573c635363.png)

**Conclusion**
Latency is significantly lower for list and inspect operations under docker 1.11.1, but CPU is much spikier (smoothed usage may be similar to 1.9.1). I will follow up with results including docker-containerd usage.

**Comment by yujuhong** [comment]
> Latency is significantly lower for list and inspect operations under docker 1.11.1, but CPU is much spikier (smoothed usage may be similar to 1.9.1). I will follow up with results including docker-containerd usage.

If it helps, I remember the docker-containerd cpu usage to be quite low that it's insignificant.

**Comment by timstclair** [comment]
Yeah, that's consistent with my findings. I've updated the graphs to include containerd.

**Comment by Random-Liu** [comment]
Maybe the CPU spike is because docker processes request much faster than before, the benchmark sends more request in the same period.

**Comment by yujuhong** [comment]
We should check whether https://github.com/kubernetes/kubernetes/issues/21751#issuecomment-189552291 was fixed in v1.11
Basically, running a bunch of containers and run `exec` and `stop` frequently to see if `stop` hangs.

**Comment by spiffxp** [comment]
What remains to be done here?  Is 1.11 going to be the recommendation for 1.3?

**Comment by timstclair** [comment]
We're still doing more testing. https://github.com/kubernetes/kubernetes/issues/25844 is the current blocker for switching our master branch to docker v1.11 for soak testing.

**Comment by dchen1107** [comment]
#25844 was merged and reverted shortly because it breaks kubernetes-kubemark-500-gce (#25949). We are still investigating the root cause. One possible culprit pointed by @yujuhong is that it seems to be caused by the fact that runc creates a unique session key for each container (opencontainers/runc#726). If that is the root cause, the max-number of containers created by docker is capped by /proc/sys/kernel/keys/maxkeys, and I checked the containervm image, the default value is 200. The puzzle is that we tested it by starting 500 pause containers on the node, it works without capping.

**Comment by dchen1107** [comment]
I reproduced the issue on my test KNode, and filed one against docker at: https://github.com/docker/docker/issues/22865 

I am still wondering why we couldn't reproduce the problem on @timstclair's node. The only difference on his node is that his node doesn't run kubelet.

**Comment by vishh** [comment]
Can we just set a higher limit in kubelet for now?

**Comment by yujuhong** [comment]
I also reproduced this with @dchen1107's command on my node.
Bumping `/proc/sys/kernel/keys/root_maxkeys` worked for me, until the ip addresses ran out.

**Comment by dchen1107** [comment]
@vishh and @yujuhong bumping the limit is the plan I am doing. :-) How about 2000?

**Comment by dchen1107** [comment]
Meanwhile, @timstclair is hitting other issue related to cAdvisor and docker 1.11.1. He is still reproducing that one. I will send my fix shortly.

**Comment by vishh** [comment]
We might want to bump `/proc/sys/kernel/keys/root_maxbytes` too. Seems to
be set to `20000` bytes by default.

**Comment by timstclair** [comment]
I was able to reproduce this on a container VM GCE instance, but not on an ubuntu trusty instance (that I installed docker v1.11 on using `curl -fsSL https://get.docker.com/ | sh`)

**Comment by Random-Liu** [comment]
It's wried that docker micro benchmark https://github.com/kubernetes/kubernetes/issues/23397#issuecomment-217055068 didn't catch the error https://github.com/kubernetes/kubernetes/issues/23397#issuecomment-220692094.

I think one reason is old version of docker micro benchmark didn't report docker operation errors, maybe 
we've already hit the error during the test.
https://github.com/kubernetes/contrib/pull/581 will add error rate statistics. I'll run new docker micro benchmark with 1.11 again.

**Comment by yujuhong** [comment]
@Random-Liu, 200 running containers won't hit this case. In addition to that, one needs to run the benchmark using a containerVM instance to hit this case (or it depends on the settings on your desktop). We should get your PR in to at least surface the errors when running the microbenchmark

**Comment by dchen1107** [comment]
@Random-Liu and I talked about this last week, and we are extending the test suite for this.

**Comment by Random-Liu** [comment]
@dchen1107 ACK. Will increase the max container number in benchmark test.

**Comment by vishh** [comment]
@timstclair https://github.com/kubernetes/kubernetes/issues/23397#issuecomment-209583923 is still unresolved AFAIK.

**Comment by timstclair** [comment]
Ah, I missed that. Thanks.

**Comment by yujuhong** [comment]
Docker 1.11.2 is out: https://github.com/docker/docker/releases/tag/v1.11.2

Fixes that may be relevant:
 - Fix a possible deadlock on image deletion and container attach
 - Fix an issue preventing container to be restarted after daemon restart 
 - Fix an issue causing docker ps to hang on CentOS when using devicemapper (#22168, #23067)
 - Fix a bug preventing to docker exec into a container when using devicemapper (#22168, #23067)

**Comment by dchen1107** [comment]
@thockin reported https://github.com/docker/containerd/issues/275

**Comment by bgrant0607** [comment]
@timstclair Is there anything left to do here?

**Comment by timstclair** [comment]
Nope, we've been defaulting to v1.11 on GCE / GKE since 1.3.

**Comment by dchen1107** [comment]
Known issue of docker 1.11.X: https://github.com/kubernetes/kubernetes/issues/33882#issuecomment-264075001 by @mtaufen 

This is theorized to be the fix in Docker: docker/docker#25523
That fix goes in with Docker 1.13.0.
